signac/contrib/utility.py

```diff
@@ -479,3 +481,255 @@ def _nested_dicts_to_dotted_keys(t, encode=_encode_tree, key=None):
+    exclude_keys : sequence[str]
+        A sequence of keys to exclude (Default value = None).
+    delimiter_nested : str
+        Delimiter used for nesting keys (Default value = '.').
+
+    Returns
+    -------
+    callable
+        Function that returns a normalized path.
+
+    """
+    from .schema import _build_job_statepoint_index
+
+    if len(jobs) <= 1:
+        # The lambda must (optionally) take a format spec argument to match the
+        # signature of the path function below.
+        return lambda job, sep=None: ""
+
+    index = [{"_id": job.id, "sp": job.sp()} for job in jobs]
+    jsi = _build_job_statepoint_index(exclude_const=True, index=index)
+    sp_index = OrderedDict(jsi)
+
+    paths = {}
+    for key_tokens, values in sp_index.items():
+        key = key_tokens.replace(".", delimiter_nested)
+        if exclude_keys and key in exclude_keys:
+            continue
+        for value, group in values.items():
+            path_tokens = key, str(value)
+            for job_id in group:
+                paths.setdefault(job_id, list())
+                paths[job_id].extend(path_tokens)
+
+    def path(job, sep=None):
+        """Normalize the path.
+
+        Parameters
+        ----------
+        job : :class:`~signac.contrib.job.Job`
+            An instance of :class:`~signac.contrib.job.Job`.
+        sep : str
+            (Default value = None)
+
+        Returns
+        -------
+        str
+            Normalized path.
+
+        Raises
+        ------
+        RuntimeError
+            If unable to determine path for job.
+
+        """
+        try:
+            if sep:
+                return os.path.normpath(sep.join(paths[job.id]))
+            else:
+                return os.path.normpath(os.path.join(*paths[job.id]))
+        except KeyError:
+            raise RuntimeError(
+                "Unable to determine path for job '{}'.\nThis is usually caused by a "
+                "heterogeneous schema, where some keys are only present in some jobs. "
+                "Try providing a custom path.".format(job)
+            )
+
+    return path
+
+
+class _AutoPathFormatter(Formatter):
+    """A custom formatter used to format jobs with their respective paths."""
+
+    def __init__(self, paths):
+        self.paths = paths
+
+    def format_field(self, value, format_spec):
+        """Format string fields, replacing jobs with their paths.
+
+        Parameters
+        ----------
+        value :
+            The value to be formatted.
+        format_spec :
+            The format specification.
+
+        Returns
+        -------
+        str
+            Formatted string.
+
+        """
+        from .job import Job
+
+        if isinstance(value, Job):
+            return self.paths(value, format_spec)
+        else:
+            return super().format_field(value, format_spec)
+
+
+class _SchemaPathEvaluationError(RuntimeError):
+    """Raised for errors in schema path evaluation."""
+
+    pass
```

Comment:
This needs to be a public class (no underscore) defined in `signac/contrib/errors.py`, because it can be raised to the user.